src/Microsoft.Azure.SignalR/ClientInvocation/CallerClientResultsManager.cs

```diff
@@ -0,0 +1,195 @@
+                if (invocation.RouterInstanceId == instanceId)
+                {
+                    var message = new CompletionMessage(invocationId, $"Connection '{invocation.ConnectionId}' is disconnected.", null, false);
+                    
+                    invocation.Complete(invocation.Tcs, message);
+                    _pendingInvocations.TryRemove(invocationId, out _);
+                }
+            }
+        }
+
+        public bool TryCompleteResult(string connectionId, CompletionMessage message)
+        {
+            if (_pendingInvocations.TryGetValue(message.InvocationId, out var item))
+            {
+                if (item.ConnectionId != connectionId)
+                {
+                    throw new InvalidOperationException($"Connection ID '{connectionId}' is not valid for invocation ID '{message.InvocationId}'.");
+                }
+
+                // if false the connection disconnected right after the above TryGetValue
+                // or someone else completed the invocation (likely a bad client)
+                // we'll ignore both cases
+                if (_pendingInvocations.TryRemove(message.InvocationId, out _))
+                {
+                    item.Complete(item.Tcs, message);
+                    RemoveServiceMapping(message.InvocationId);
+                    return true;
+                }
+                return false;
+            }
+            else
+            {
+                // connection was disconnected or someone else completed the invocation
+                return false;
+            }
+        }
+
+        public bool TryCompleteResult(string connectionId, ClientCompletionMessage message)
+        {
+            var protocol = _hubProtocolResolver.GetProtocol(message.Protocol, null);
+            if (protocol == null)
+            {
+                throw new InvalidOperationException($"Not supported protocol {message.Protocol} by server");
+            }
+
+            var payload = message.Payload;
+            if (protocol.TryParseMessage(ref payload, this, out var hubMessage))
+            {
+                if (hubMessage is CompletionMessage completionMessage)
+                {
+                    return TryCompleteResult(connectionId, completionMessage);
+                }
+                else
+                {
+                     throw new InvalidOperationException($"The payload of ClientCompletionMessage whose type is {hubMessage.GetType()} cannot be parsed into CompletionMessage correctly.");
+                }
+            }
+            return false;
+        }
+
+        // Implemented for interface IInvocationBinder
+        public Type GetReturnType(string invocationId)
+        {
+            if (TryGetInvocationReturnType(invocationId, out var type))
+            {
+                return type;
+            }
+            throw new InvalidOperationException($"Invocation ID '{invocationId}' is not associated with a pending client result.");
```

Comment:
Is it possible that invocation is already removed? What happens to the customer when it throws?